### PR TITLE
Tune vectorized binary_search, lower_bound, and upper_bound algorithms

### PR DESCRIPTION
**Pull request by mmichel11**
This PR introduces some additional tuning for the vectorized search algorithms (binary_search, lower_bound, and upper_bound) where a series of searches across a provided set of search keys are performed on a single input buffer.

The improvements are explained below:

1. In the current implementations, `__pstl_lower_bound`  / `__pstl_upper_bound` have been passed a 64-bit size type to compute indicies in its binary search. For inputs with a large number of search keys, this is a bottleneck in performance and rooflines have shown that this kernel is bound by 64-bit arithmetic. This PR introduces a 2-kernel solution where a 32-bit unsigned type is used for index calculation if there are less than ``2^32`` search keys with a fallback to 64-bit offset calculation otherwise. This results in a performance improvement across these algorithms of `~1.3x` on GPU Series Max 1550 for large inputs.
2. The second improvement is in the implementation of `__pstl_lower_bound`. Simplifying the loop body by checking descending powers-of-2 as opposed to continuously dividing the input size by `2` reduces the amount of total arithmetic in the compiled code and results in a performance improvement of `10-15%` for large inputs on top of the changes in `1`. This improvement requires the size type to be unsigned and is implemented as a new utility function: `__shars_lower_bound`.

### CODE REVIEW

**Comment by SergeyKopienko** [comment]
@mmichel11, let's merge this PR only after https://github.com/oneapi-src/oneDPL/pull/1239, ok?

**Comment by mmichel11** [comment]
> @mmichel11, let's merge this PR only after #1239, ok?

Sounds good. I will not merge it until after the tag dispatching PR and rebase.

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/utils.h:726]: const ?

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/internal/binary_search_impl.h:241]: Do we really need to have `fallback_policy` as local variable?

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/internal/binary_search_impl.h:138]: The code in these two branches looks like the same excepting used policy.
May we combine it into one branch?

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/internal/binary_search_impl.h:185]: The same

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/internal/binary_search_impl.h:233]: The same

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h:138]: I don't think it is possible here. In the first branch `custom_brick` uses a `::std::uint32_t` index type and a `::std::size_t` in the second branch.  We have to make the dispatch decision at runtime to choose the correct kernel.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (danhoeflinger) [inline include/oneapi/dpl/internal/binary_search_impl.h:190]: Seems like we can reduce this to a lambda and make the branches just provide `policy` and `size` of the appropriate type.

It adds the complexity of the lambda to reduce the code replication, so its not a clear win, but I think it would be better.
(same comment for all three instances)

File: include/oneapi/dpl/pstl/utils.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/utils.h:750]: should this use `__shars_lower_bound` rather than `__pstl_lower_bound`?

**Review by julianmi — COMMENTED** [review]
These changes make sense to me. There might be potential benefits of utilizing 32-bit addressing on other algorithms as well. This is beyond this PR but maybe something to keep in mind.

**Comment by danhoeflinger** [comment]
> There might be potential benefits of utilizing 32-bit addressing on other algorithms as well. 

I agree with this, but I also feel like this may be something to take up with the `sycl` runtime folks.  Its not a good user experience to need to have this sort of logic to achieve good performance (and double the number of compiled kernels).  Of course, we do not want to have the runtime automatically double the number of kernels for us, but they may be able to improve the performance of the 64-bit version somehow.

Another option @mmichel11 and I were brainstorming offline:

Adding some opt-in define which would use 32-bit addressing everywhere, where the user guarantees they wont be calling the algorithms for `n > max(uint32_t)`, and in return gets a performance boost across the board.

**Comment by julianmi** [comment]
> > There might be potential benefits of utilizing 32-bit addressing on other algorithms as well.
> 
> I agree with this, but I also feel like this may be something to take up with the `sycl` runtime folks. Its not a good user experience to need to have this sort of logic to achieve good performance (and double the number of compiled kernels). Of course, we do not want to have the runtime automatically double the number of kernels for us, but they may be able to improve the performance of the 64-bit version somehow.
> 
> Another option @mmichel11 and I were brainstorming offline:
> 
> Adding some opt-in define which would use 32-bit addressing everywhere, where the user guarantees they wont be calling the algorithms for `n > max(uint32_t)`, and in return gets a performance boost across the board.

I agree with this and especially the opt-in define. Many of our size-related types are derived from SYCL's `size()` or previously `get_count()` which seem to use 64-bits. We could cast them down if the user permits. We should probably investigate the performance benefits in more detail.

File: include/oneapi/dpl/pstl/utils.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h]: Yes, I have made this change.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h]: Changed for all three instances to use a lambda.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h]: No, I have changed this.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h]: Changed

File: include/oneapi/dpl/pstl/utils.h
Comment (adamfidel) [inline include/oneapi/dpl/pstl/utils.h:722]: Is there a reason to use the name `_Size1` instead of `_Size`? I don't see a `_Size2`.

Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h]: No, I just followed what was being done in `__pstl_lower_bound`. It is confusing, so I have changed it to `_Size`.

**Comment by mmichel11** [comment]
> > > There might be potential benefits of utilizing 32-bit addressing on other algorithms as well.
> > 
> > 
> > I agree with this, but I also feel like this may be something to take up with the `sycl` runtime folks. Its not a good user experience to need to have this sort of logic to achieve good performance (and double the number of compiled kernels). Of course, we do not want to have the runtime automatically double the number of kernels for us, but they may be able to improve the performance of the 64-bit version somehow.
> > Another option @mmichel11 and I were brainstorming offline:
> > Adding some opt-in define which would use 32-bit addressing everywhere, where the user guarantees they wont be calling the algorithms for `n > max(uint32_t)`, and in return gets a performance boost across the board.
> 
> I agree with this and especially the opt-in define. Many of our size-related types are derived from SYCL's `size()` or previously `get_count()` which seem to use 64-bits. We could cast them down if the user permits. We should probably investigate the performance benefits in more detail.

This is an investigation I plan to do in the future. There are several places now where we define separate kernels to take advantage of 32-bit index types when possible. I think an opt-in define would be the best solution instead of doubling the number of kernels per algorithm.

**Comment by mmichel11** [comment]
After offline discussion with @danhoeflinger, we agreed it would be best to group the upper / lower bound implementations together. I have moved them after the bit algorithms as they are required to implement `__shars_lower_bound`.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (akukanov) [inline include/oneapi/dpl/internal/binary_search_impl.h:157]: Have you tried to move the runtime decision to `custom_brick` instead? As far as I can see, the size value to make the decision is the same on the host and in the brick, so that should be technically possible. How would that perform?

Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h:157]: I did some experimenting and there seems to be a small performance penalty to performing the check within the kernel itself (<1% for large n and large number of search keys). However, this allows us to have a compiled single kernel which is likely worth it.

I will do a bit more investigation of doing this runtime check in the kernel.

File: include/oneapi/dpl/pstl/utils.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h]: Missed responding to this. I made this variable `const` as well as others where applicable.

**Comment by mmichel11** [comment]
@julianmi @SergeyKopienko @akukanov @danhoeflinger @adamfidel

Are there any additional comments / concerns you have for this PR, or are you ready to approve?

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/utils.h:693]: 1) Are we able to rewrite this like `__search_offset += __cur_pow2;` ?
2) What about remove this `if` ? 
```C++
__search_offset  += __comp(__acc[__search_idx], __value) ? __cur_pow2 : 0;
```

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [inline include/oneapi/dpl/pstl/utils.h:666]: Revere counter iterator looks like a little bit strange for me,
As far as I know we already have `oneapi::dpl::counting_iterator`. Is it not enough?
Also when I see the API like `_Index __first, _Index __last` I assume that the element before `__first` isn't exist and the element iterated by `__last` isn't exist too. From this point of view what does it mean `__first - 1` below?

@danhoeflinger what do you think about this?

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/binary_search_impl.h]: As described at https://en.cppreference.com/w/cpp/types/size_t, "The bit width of std::size_t is not less than 16.(since C++11)"
So does this solution quite correct when `size > ::std::numeric_limits<::std::uint32_t>::max()` ?
Is OK that n this case you simple using `static_cast<::std::size_t>(size)` ?
May be we should use some another type instead of `std::size_t` ?

File: include/oneapi/dpl/pstl/utils.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h:693]: For 1, using bitwise or proved to be slightly faster, so I think sticking with that is best.

I should be able to use your suggestion from 2 still. Let me try check performance with that change.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (mmichel11) [reply include/oneapi/dpl/internal/binary_search_impl.h]: I see your point here. I had used `size_t` for the opposite case where it could holder larger values than `uint64_t`. In other places in the code where we do this, I see we use `std::uint64_t` as the fallback. 

Does switching to `std::uint64_t` seem like a good alternative to you?

File: include/oneapi/dpl/pstl/utils.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h:693]: I tried the approach you suggested with the ternary operator and a "no-op" bitwise or with 0 in the case the predicate is not satisfied. Interestingly, this ended up being slightly slower. It seems that the compiler is able to better optimize with the way it is currently written, so my thoughts are we leave it as is.

File: include/oneapi/dpl/pstl/utils.h
Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h:666]: @SergeyKopienko and I discussed offline and agreed. This code is just being moved so that we can group these similar utilities together.

If there is an issue or room for improvement in `__pstl_left_bound`, then we can open a GH issue to address in a future PR.

File: include/oneapi/dpl/internal/binary_search_impl.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/internal/binary_search_impl.h]: Probably yes.

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h:693]: ok

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h:666]: So, @mmichel11, could you please create the issue for this point?

**Comment by SergeyKopienko** [comment]
@mmichel11 no any comments or concerns from my side. I am ready to approve this PR because I can't find something incorrect anymore. But please receive additional approve from somebody too,

File: include/oneapi/dpl/pstl/utils.h
Comment (danhoeflinger) [inline include/oneapi/dpl/pstl/utils.h:689]: My assumption is that the compiler will always replace these divisions by two with shift right by one.  You've probably checked performance comparison, if not it could be worth a look.

So if they are the same performance, then we have to decide which is more clear to read.  My preference for the ones in this loop would be to use `>> 1`, since we are effectively setting each bit of the index with each loop iteration, and we use bitwise ops elsewhere in the loop.

Just my preference, feel free to leave as is.

**Review by danhoeflinger — APPROVED** [review]
LGTM, optional change if you want to apply I'll reapprove.

File: include/oneapi/dpl/pstl/utils.h
Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h:689]: I read that `operator>>' isn't so clear as we thinking for the signed types with the value `< 0`. The interesting moment is the value of sign bit and highest bit(s) which state should be assigned during the `operator>>' implementation. See https://stackoverflow.com/questions/141525/what-are-bitwise-shift-bit-shift-operators-and-how-do-they-work for details. So would prefer `/2` operation when the type is unknown.

Comment (danhoeflinger) [reply include/oneapi/dpl/pstl/utils.h:689]: I agree with you for unknown types, but this function has a static_assert enforcing that `_Size` is an unsigned type.

Comment (SergeyKopienko) [reply include/oneapi/dpl/pstl/utils.h:689]: So it's OK in this case.

Comment (mmichel11) [reply include/oneapi/dpl/pstl/utils.h]: Performance is identical between these two versions, and unsigned types are required for the bit floor call.

I wasn't sure which one is clearer when writing this, but I think `>>=` makes the most sense since we use other bitwise operations in the loop.

I have went ahead and made this change.

**Review by danhoeflinger — APPROVED** [review]
LGTM

**Review by julianmi — APPROVED** [review]
LGTM

**Pull request merged**
